### PR TITLE
draft: Cloudflare works ! 🎉 

### PR DESCRIPTION
**Pull request by wackfx**
Hey there !

Amazing work with your lib. It's so awesome that I was really sad to see that it didn't work on cloudflare 😢 

I've spent some time to make it work on Cloudflare. Tested on my own project - it's working for both Cloudflare pages workers and regular workers.

This PR is **not** ready to merge - but since I've spent hours trying to deploy this library on Cloudflare based on your work - I guess you may want to take some pointers to what's missing from your implementation to successfully deploy it.

### Here's what changed:
- Rebased your current work on `master`
- polyfill: `Crypto.subtle.digest(MD5)` didn't work as md5 is not supported anymore. Added Joseph's Myers implementation to follow your pattern of 'no external libs'
- polyfill: Added dynamic import of `cloudflare:socket` to please bundlers
- transpile: Replacing `net.Socket()` with `async net.Socket()` to allow for dynamic import on CF code
- polyfill: `os` `fs` polyfills that doesn't respect the original contract - but since we're on cloudflare there is not much more we can do
- polyfill: `process.hrtime`
- transpile: Buffer being undefined on cloudflare - added import on build:cf
- transpile: added `node:` prefixes

### How to use ?

Simply `import postgres from 'postgres'` and instantiate it like you would do in a regular node env.

I use this module with Qwik and the adapter cloudflare-pages.
You'll need to tell bundler (through vite.config.ts on Qwik) to resolve `workerd` and set the node library (but compatible with cf workers) as external like this:
image

That's it. Feel free to take whatever you want and close this PR, I won't take it personally, I know it's pretty hacky ;)
Thanks again for your initial work.

### CODE REVIEW

**Comment by porsager** [comment]
Oh my- very very cool, but did you not see https://github.com/porsager/postgres/pull/599 ? 😬

**Comment by wackfx** [comment]
Oh yes, I saw that, the `cloudflare` branch was actually the base work for this (your commits are in my branch!), but it didn't work when deployed on Cloudflare 😞

**Comment by porsager** [comment]
Did you use node-compat? The CloudFlare workers team tested that branch with successOn 24 Jun 2023, at 10.44, Wack ***@***.***> wrote:﻿
Oh yes, I saw that, the cloudflare branch was actually the base work for this (your commits are in my branch!), but it didn't work when deployed on Cloudflare 😞

**Comment by porsager** [comment]
Ok, back at computer. I should have read your entire description, I just didn't had time when I answered ;) I'll look closer later or tomorrow.

**Comment by porsager** [comment]
Even so, still curious if you tried with node-compat?

**Comment by wackfx** [comment]
Hey, sorry I missed the notification !

Yes, I did try it to run it with wranger and `node_compat`, with a simple JS test like this:

```javascript
import postgres from './cf/src/index'
const DB_URI = "..." // URI
const sql = postgres(DB_URI)
export function onRequest(context) {
  console.log(await sql`SELECT * FROM users`)
  return new Response("Hello, world!")
}
```

Unfortunately, tried on both Railway and Supabase, it didn't work. It blocked at the authentication.

I did solve Railway by patching the `md5` function (because the authentication sent by the server was a standard unprotected one)

Later on, I solved Supabase by patching the `sha256` method because it asked for a SASL authentication.

Main blocking issue I've found was because of data types: digest would wait for a Uint8Array (from `decode/encode`) and I was giving it a string - which made the whole thing freeze. Also, the `connection.js` logic required to get back an ArrayBuffer (and avoid any other type) to be functional.

Also, Cloudflare complained once I tried to deploy that some `globals` or modules was missing (because polyfill wouldn't add them). You really have to try to deploy something to make sure it works: sometimes it worked perfectly locally, but then Cloudflare would complain at deploy time.

I've my project rn running on Cloudflare Pages (with some workers functions) - I can give you the link if you want 😄

**Comment by porsager** [comment]
Very cool you're getting these last things in order. I was supposed to merge quite some time ago, but haven't had the time.

Mind rebasing your changes on top of my cloudflare branch to get a better view of what changed? (just rebased my cloudflare branch on master now)

**Comment by wackfx** [comment]
Hey !

I've rebased on top of your clourflare branch - sorry about the formatting issues.

Let me know if I can be of any more help 🙂 

Thank you

**Comment by porsager** [comment]
I don't know why it still complains. Mind resolving the conflicts, and removing the md5 implementation?

**Comment by daniel-xyz** [comment]
That's awesome. Hopefully that means it'll soon work on Vercel Edge Functions too, since they're built on top of Cloudflare Workers as far as I know. 👀

**Comment by wackfx** [comment]
I believe @superboss224 had the same issues that I've fixed (looking at the import errors he gave). Definitely a disprecancy between Cloudflare Worker's and Page's workers.
Although depending on the database he uses, he may need some tinkering with the code.

**Comment by wackfx** [comment]
Sorry I'm not on my computer right now to fix the issues. May be able to take a look tomorrow. I've got this code running actually on Cloudflare Pages with Qwik ([https://shwink.cloudflare.com](here)) so I do believe it's working.

**Comment by superboss224** [comment]
> I believe @superboss224 had the same issues that I've fixed (looking at the import errors he gave). Definitely a disprecancy between Cloudflare Worker's and Page's workers. Although depending on the database he uses, he may need some tinkering with the code.

Yes exactly !
Cloudflare Workers has a more powerful `node_compat = true` flag that also includes polyfills for native module, **but** that flag is getting more or less legacy, and I don't think that Cloudflare wants to continue supporting it actively in the future.

They prefer the new `compatibility_flags = [ "nodejs_compat" ]`

More info about the two different flags here :
https://developers.cloudflare.com/workers/wrangler/configuration/#add-polyfills-using-wrangler

The difference in the doc may seem subtle, but it is what is causing the discrepancy between Worker and Pages 

(I'm not sure that the new _nodejs_compat_ is necessary if everything can be polyfilled, but that new flag should work in Pages

**Comment by superboss224** [comment]
Yep, giga fast, there's a recent switch to push to that truly "javascript edge runtime" that has nothing to do with node.js or native modules (on Vercel Edge, Cloudflare and other platforms)

But we can now use a real tcp connexion in those runtimes and it's awesome because they are closer to the end user and have almost no cold start (meanwhile Vercel Functions has 2-3 seconds cold start and an user will **feel** it)

And being able to use a DBs in those edge runtimes was the last piece missing to run entire Full Stacks apps on it and avoid downtimes/scaling issues/latency for good, so having Postgres.js working with it will be awesome !

**Comment by porsager** [comment]
@wackfx been looking at this, and was quite hard to diff with the formatting changes. I've updated my branch with the things I could glance was necessary from your changes. Do you see anything missing?

File: cf/polyfills.js [hunk 1]
```diff
@@ -13,61 +13,69 @@ const IPv4Reg = new RegExp(`^${v4Str}$`)
...
-    `(?:${v6Seg}:){7}(?:${v6Seg}|:)|` +
-    `(?:${v6Seg}:){6}(?:${v4Str}|:${v6Seg}|:)|` +
-    `(?:${v6Seg}:){5}(?::${v4Str}|(:${v6Seg}){1,2}|:)|` +
-    `(?:${v6Seg}:){4}(?:(:${v6Seg}){0,1}:${v4Str}|(:${v6Seg}){1,3}|:)|` +
-    `(?:${v6Seg}:){3}(?:(:${v6Seg}){0,2}:${v4Str}|(:${v6Seg}){1,4}|:)|` +
-    `(?:${v6Seg}:){2}(?:(:${v6Seg}){0,3}:${v4Str}|(:${v6Seg}){1,5}|:)|` +
-    `(?:${v6Seg}:){1}(?:(:${v6Seg}){0,4}:${v4Str}|(:${v6Seg}){1,6}|:)|` +
-    `(?::((?::${v6Seg}){0,5}:${v4Str}|(?::${v6Seg}){1,7}|:))` +
-    ')(%[0-9a-zA-Z-.:]{1,})?$'
+  `(?:${v6Seg}:){7}(?:${v6Seg}|:)|` +
```
Comment (wackfx) [inline cf/polyfills.js:16]: Just some linting issues - I've went ahead and used `eslint` on this file but no kudos :(

File: cf/polyfills.js [hunk 2]
```diff
@@ -13,61 +13,69 @@ const IPv4Reg = new RegExp(`^${v4Str}$`)
...
+        iterations
+      },
+      await Crypto.subtle.importKey(
+        'raw',
+        textEncoder.encode(password),
+        'PBKDF2',
+        false,
+        ['deriveBits']
+      ),
+      keylen * 8,
```
Comment (wackfx) [inline cf/polyfills.js:45]: Nothing changed here

File: cf/polyfills.js [hunk 3]
```diff
@@ -13,61 +13,69 @@ const IPv4Reg = new RegExp(`^${v4Str}$`)
...
-    keylen * 8,
-    ['deriveBits']
-  ),
   createHash: (type) => ({
     update: (x) => ({
       digest: () => {
-        return type === 'sha256'
-          ? Crypto.subtle.digest('SHA-256', x)
-          : Crypto.subtle.digest('MD5', x)
+        if (type !== 'sha256')
```
Comment (wackfx) [inline cf/polyfills.js:51]: Removed MD5 as you asked

File: cf/polyfills.js [hunk 4]
```diff
@@ -13,61 +13,69 @@ const IPv4Reg = new RegExp(`^${v4Str}$`)
...
     })
   }),
   createHmac: (type, key) => ({
-    update: x => ({
-      digest: async() => Buffer.from(await Crypto.subtle.sign(
-        'HMAC',
-        await Crypto.subtle.importKey('raw', key, { name: 'HMAC', hash: 'SHA-256' }, false, ['sign']),
-        textEncoder.encode(x)
-      ))
+    update: (x) => ({
```
Comment (wackfx) [inline cf/polyfills.js:60]: Linting only

File: cf/polyfills.js [hunk 5]
```diff
@@ -13,61 +13,69 @@ const IPv4Reg = new RegExp(`^${v4Str}$`)
...
+              ['sign']
+            ),
+            textEncoder.encode(x)
+          )
+        )
     })
   })
 }
 
-export const process = {
```
Comment (wackfx) [inline cf/polyfills.js:67]: added at the end of the file because of `hrtime` polyfill

File: cf/polyfills.js [hunk 6]
```diff
@@ -81,19 +89,23 @@ export const fs = {
 }
 
 export const net = {
-  isIP: x => RegExp.prototype.test.call(IPv4Reg, x) ? 4 : RegExp.prototype.test.call(IPv6Reg, x) ? 6 : 0,
+  isIP: (x) =>
```
Comment (wackfx) [inline cf/polyfills.js:92]: linting only

File: cf/polyfills.js [hunk 7]
```diff
@@ -81,19 +89,23 @@ export const fs = {
...
+        : 0,
   Socket
 }
 
 export { setImmediate, clearImmediate }
 
 export const tls = {
-  connect(x) {
-    const tcp = x.socket
+  connect({ socket: tcp, servername }) {
```
Comment (wackfx) [inline cf/polyfills.js:104]: Destructuring here 🤔 
Not sure if you want to keep this, maybe a merge side-effect. Feel free to keep it or not - this code works (I've added some tests files down this PR)

File: cf/polyfills.js [hunk 8]
```diff
@@ -133,7 +145,7 @@ function Socket() {
         () => {
           tcp.readyState !== 'upgrade'
             ? close()
-            : (tcp.readyState = 'open', tcp.emit('secureConnect'))
+            : ((tcp.readyState = 'open'), tcp.emit('secureConnect'))
```
Comment (wackfx) [inline cf/polyfills.js:148]: linting

File: cf/polyfills.js [hunk 9]
```diff
@@ -151,8 +163,7 @@ function Socket() {
   }
 
   function close() {
-    if (tcp.readyState === 'closed')
-      return
+    if (tcp.readyState === 'closed') return
```
Comment (wackfx) [inline cf/polyfills.js:166]: linting

File: cf/polyfills.js [hunk 10]
```diff
@@ -164,9 +175,7 @@ function Socket() {
   }
 
   function end(data) {
-    return data
-      ? tcp.write(data, () => tcp.raw.close())
-      : tcp.raw.close()
+    return data ? tcp.write(data, () => tcp.raw.close()) : tcp.raw.close()
```
Comment (wackfx) [inline cf/polyfills.js:178]: linting

File: cf/polyfills.js [hunk 11]
```diff
@@ -178,7 +187,7 @@ function Socket() {
     try {
       let done
         , value
-      while (({ done, value } = await tcp.reader.read(), !done))
+      while ((({ done, value } = await tcp.reader.read()), !done))
```
Comment (wackfx) [inline cf/polyfills.js:190]: linting

File: cf/polyfills.js [hunk 12]
```diff
@@ -211,3 +220,31 @@ function setImmediate(fn) {
 function clearImmediate(id) {
   tasks.delete(id)
 }
+
+const nowOffset = Date.now()
```
Comment (wackfx) [inline cf/polyfills.js:224]: Added `hrtime` polyfill because it was used in the code (as `process.hrtime`)

File: cf/src/connection.js [hunk 13]
```diff
@@ -1,9 +1,5 @@
-import { process } from '../polyfills.js'
 import { Buffer } from 'node:buffer'
-import { setImmediate, clearImmediate } from '../polyfills.js'
-import { net } from '../polyfills.js'
-import { tls } from '../polyfills.js'
-import { crypto } from '../polyfills.js'
+import { setImmediate, clearImmediate, process, net, tls, crypto } from '../polyfills.js'
```
Comment (wackfx) [inline cf/src/connection.js:2]: Reworked `polyfill.cf.js` to allow for a single `../polyfills.js` import (eslint was not happy about the multiple imports)

File: cf/src/index.js [hunk 14]
```diff
@@ -1,6 +1,4 @@
-import { process } from '../polyfills.js'
-import { os } from '../polyfills.js'
-import { fs } from '../polyfills.js'
+import { process, os, fs } from '../polyfills.js'
```
Comment (wackfx) [inline cf/src/index.js:1]: same here

File: cf/test-pages.js [hunk 15]
```diff
@@ -0,0 +1,22 @@
+// Add your database url and run this file with
```
Comment (wackfx) [inline cf/test-pages.js:1]: This file gives some instructions to test the code on pages wrangler (notice the `pages` in the command)

File: cf/test-worker.js [hunk 16]
```diff
@@ -0,0 +1,15 @@
+// Add your database url and run this file with
```
Comment (wackfx) [inline cf/test-worker.js:1]: Regular wrangler test

File: transpile.cf.js [hunk 17]
```diff
@@ -1,38 +1,47 @@
 import fs from 'fs'
 import path from 'path'
 
-const empty = x => fs.readdirSync(x).forEach(f => fs.unlinkSync(path.join(x, f)))
-    , ensureEmpty = x => !fs.existsSync(x) ? fs.mkdirSync(x) : empty(x)
+const empty = (x) =>
```
Comment (wackfx) [inline transpile.cf.js:4]: linter

File: transpile.cf.js [hunk 18]
```diff
@@ -1,38 +1,47 @@
...
+      fs.readdirSync(x).forEach((f) => fs.unlinkSync(path.join(x, f)))
+    , ensureEmpty = (x) => (!fs.existsSync(x) ? fs.mkdirSync(x) : empty(x))
     , root = 'cf'
     , src = path.join(root, 'src')
 
 ensureEmpty(src)
 
-fs.readdirSync('src').forEach(name =>
+fs.readdirSync('src').forEach((name) =>
```
Comment (wackfx) [inline transpile.cf.js:12]: linter

File: transpile.cf.js [hunk 19]
```diff
@@ -1,38 +1,47 @@
...
 
 function transpile(x) {
-  const timers = x.includes('setImmediate')
-    ? 'import { setImmediate, clearImmediate } from \'../polyfills.js\'\n'
-    : ''
-
-  const process = x.includes('process.')
-    ? 'import { process } from \'../polyfills.js\'\n'
-    : ''
+  const polyfills = [
```
Comment (wackfx) [inline transpile.cf.js:20]: Reworked polyfill.cf.js to allow for a single ../polyfills.js import (eslint was not happy about the multiple imports)

**Review by wackfx — COMMENTED** [review]
Added some PR comments

**Comment by wackfx** [comment]
Hey @porsager - I've made the latest changes / rebased on your cloudflare work.
I don't belive there is much changing, besides the `transpile.cf.js` rework to make eslint happy + hrtime polyfill.
Also added a couple of tests files for you to test cloudflare (pages + worker)

Feel free to take whatever you need.
Best regards

File: transpile.cf.js (hunk 17 above)
Comment (porsager) [reply transpile.cf.js:4]: Are you sure you're using the .eslint from this repo?

**Comment by porsager** [comment]
Very nice wackfx! I think the rest is just pleasing the linter. (i think you're not using this projects eslintrc file). Even so I'll just deal with that after the merge - you've done a lot already :-) Thanks!

File: cf/polyfills.js [hunk 20]
```diff
@@ -13,61 +13,69 @@ const IPv4Reg = new RegExp(`^${v4Str}$`)
...
-    ['deriveBits']
-  ),
   createHash: (type) => ({
     update: (x) => ({
       digest: () => {
-        return type === 'sha256'
-          ? Crypto.subtle.digest('SHA-256', x)
-          : Crypto.subtle.digest('MD5', x)
+        if (type !== 'sha256')
+          throw Error('createHash only supports sha256 on cloudflare.')
```
Comment (mattbishop) [inline cf/polyfills.js:52]: This won't just run on Cloudflare. I would suggest a more generic error message, like `createHash only supports sha256 in this environment.`

File: cf/polyfills.js [hunk 21]
```diff
@@ -81,19 +89,23 @@ export const fs = {
 }
 
 export const net = {
-  isIP: x => RegExp.prototype.test.call(IPv4Reg, x) ? 4 : RegExp.prototype.test.call(IPv6Reg, x) ? 6 : 0,
+  isIP: (x) =>
+    RegExp.prototype.test.call(IPv4Reg, x)
```
Comment (mattbishop) [inline cf/polyfills.js:93]: Curious, why not use the simpler `IPv4Reg.test(x)` ?

Comment (porsager) [reply cf/polyfills.js:93]: Straight up copy from the deno node polyfill 🙈 Might be something about their contexts and the RegExp instance used, but that shouldn't apply to us, so yes, let's change it to the simpler method :)

**Comment by superboss224** [comment]
Hello,
I use SvelteKit and looking again at the logs : 

```
```

```js
import { sveltekit } from '@sveltejs/kit/vite';

/** @type {import('vite').UserConfig} */
const config = {
  resolve: {
    conditions: ['workerd', 'webworker', 'worker', 'browser', 'import', 'module', 'default']
  },
  plugins: [sveltekit()],
  server: {
    port: 3000
  }
}

export default config;```

**Comment by superboss224** [comment]
Hello,
I use SvelteKit and looking again at the logs : 

```
21:15:48.711 | > Using @sveltejs/adapter-auto
-- | --
21:15:49.005 | ✘ [ERROR] Could not resolve "os"
21:15:49.005 |  
21:15:49.005 | node_modules/postgres/src/index.js:1:15:
21:15:49.005 | 1 │ import os from 'os'
```
It seems like that it doesn't want to import from _/cf/src_ folder, but rather directly from _/src_, so it can't use the polyfilled code at all, so it crashes

I tried to update my vite.config.js with export conditions :
```js
import { sveltekit } from '@sveltejs/kit/vite';

/** @type {import('vite').UserConfig} */
const config = {
  resolve: {
    conditions: ['workerd', 'webworker', 'worker', 'browser', 'import', 'module', 'default']
  },
  plugins: [sveltekit()],
  server: {
    port: 3000
  }
}

export default config;
```

However, it is still doing the same thing when building on cloudflare, and not importing from _/cf/src_

Maybe it's a bug from SvelteKit's building process, because even in local, I didn't find any way of using _/cf/src_

Is there something I could do better about that ? or another way to "force import" from _/cf/src_ without having to fork the package to a version that only have _/cf/src_ ? I tried `import postgres from 'postgres/cf/src/index.js'` but it gives an error

Maybe there is a way to tell cloudflare or vite to choose _/cf/src_ ?

It would be awesome if I could get some help to use the cloudflare version with SvelteKit,
Best regards

**Comment by jahir9991** [comment]
it's not working for Cloudflare pages

**Comment by JustMrMendez** [comment]
Still having this issue on sveltekit, is there something i've to setup?

**Comment by wackfx** [comment]
@superboss224 @jahir9991 @JustMrMendez 

According to your logs @superboss224, I believe you all have the same issues.
You need to use the `workerd` entrypoint, not the 'regular' `node_modules/postgres/src/index.js` 

You may need to specify a `resolve` key in the vite configuration, like this:
```
resolve: {
      conditions: ['workerd', 'webworker', 'worker', 'browser', 'import', 'module', 'default'],
},
```

Please refer to Vite / Sveltekit documentation, this is part of my code (not using SvelteKit) that actually works on Cloudflare Pages.

**Comment by wackfx** [comment]
@JustMrMendez 
i can see `rollupOptions` in your code, try with this:
```
    build: {
      rollupOptions: {
        external: [
          'cloudflare:sockets',
          'node:events', // not cloudflare
          'node:stream',
          'node:buffer'
        ]
   }
```

**Comment by JustMrMendez** [comment]
> external: [
>           'cloudflare:sockets',
>           'node:events', // not cloudflare
>           'node:stream',
>           'node:buffer'
>         ]

Same error

**Comment by jahir9991** [comment]
@wackfx what do you mean by worked entry point?

**Comment by wackfx** [comment]
Interestingly enough, it seems that using `npm install postgres` **doesn't** actually install the version containing the cloudflare compatible code, `./cf` folder is nowhere to be found (cc @porsager).

**I have something that works with sveltekit / cloudflare / postgres**

image

### Option 1 - Request changes on official `adapter-cloudflare` plugin

The official plugin is missing some really important options:
- Can't [override external](https://github.com/sveltejs/kit/blob/8c642cf2d63792e38e6304f840a039d4e4a55781/packages/adapter-cloudflare/index.js#L61). This is **important** for node-compatibility on Pages as you may want to use node modules available on Pages.
- Can't [override conditions](https://github.com/sveltejs/kit/blob/8c642cf2d63792e38e6304f840a039d4e4a55781/packages/adapter-cloudflare/index.js#L50C39-L50C39). Same thing, it doesn't cost that much to let advanced user the import conditions

They basically already have an [option](https://github.com/sveltejs/kit/blob/8c642cf2d63792e38e6304f840a039d4e4a55781/packages/adapter-cloudflare/index.js#L7) object that is really under used. I don't think those changes would be hard for them to accept (those are just regular build options).

for @porsager: It looks like the official module uses `worker` instead of `workerd` entrypoint. Not sure what's the standard and which project could change here, but wanted to make you notice this difference that might cause user issues.

### Option 2 - Use the custom adapter I made

Basically took the official adapter - and adapted it to be compatible with node / worker ([source code](https://github.com/wackfx/adapter-cloudflare-node)).

- Run `npm install https://github.com/porsager/postgres` to install Postgres from GitHub (and have the `cf` folder)
- Run `npm install -D https://github.com/wackfx/adapter-cloudflare-node` to install my custom adapter
- Change `svelte.config.js` to this
```suggestion
// change this line
// import adapter from '@sveltejs/adapter-cloudflare';

// with this line (only add -node at the end)
import adapter from '@sveltejs/adapter-cloudflare-node';

// It should work with the default kit configuration
/** @type {import('@sveltejs/kit').Config} */
const config = {
	kit: {
		adapter: adapter(),
	}
};
```
**Please, if you have any issue with this custom adapter, post it on my repository. Not here**

**Comment by JustMrMendez** [comment]
man, what a nightmare this has been:
```bash
 ERR_PNPM_PREPARE_PACKAGE  Failed to prepare git-hosted package fetched from
 "https://codeload.github.com/porsager/postgres/tar.gz/838c8daa89568e60161d6cee7d14c2ac26b696f1":
 postgres@3.3.5 prepublishOnly: `npm run lint`
spawn ENOENT
```

**Comment by porsager** [comment]
@JustMrMendez what did you expect playing with bleeding edge stuff? 😅 Your package manager (pnpm) shouldn't run the `prepublishOnly` script on install. It couldn't be more clearly named, but for some reason they might have thought that would be necessary when installing from github 🫠 I've moved on from pnpm myself because of annoyances like these, but if I recall correctly you can add `--ignore-scripts` to the install command to avoid any scripts running.

@wackfx I haven't published cf support to npm yet so that's why 😉

**Comment by jahir9991** [comment]
> @JustMrMendez what did you expect playing with bleeding edge stuff? 😅 Your package manager (pnpm) shouldn't run the `prepublishOnly` script on install. It couldn't be more clearly named, but for some reason they might have thought that would be necessary when installing from github 🫠 I've moved on from pnpm myself because of annoyances like these, but if I recall correctly you can add `--ignore-scripts` to the install command to avoid any scripts running.
> 
> @wackfx I haven't published cf support to npm yet so that's why 😉

@JustMrMendez you may be didn't build the master as there is no cf folder when we install postgres from npmjs

**Comment by porsager** [comment]
> you may be didn't build the master as there is no cf folder when we install postgres from npmjs

cf support is not released in the npm published version yet so you have to install from `porsager/postgres` to try it out.

**Comment by porsager** [comment]
>  It looks like the official module uses worker instead of workerd entrypoint. Not sure what's the standard and which project could change here, but wanted to make you notice this difference that might cause user issues.

@wackfx I've asked cloudflare what the recommended usage is, so let's see what they say :)

**Comment by JustMrMendez** [comment]
> @JustMrMendez what did you expect playing with bleeding edge stuff? 😅 Your package manager (pnpm) shouldn't run the `prepublishOnly` script on install. It couldn't be more clearly named, but for some reason they might have thought that would be necessary when installing from github 🫠 I've moved on from pnpm myself because of annoyances like these, but if I recall correctly you can add `--ignore-scripts` to the install command to avoid any scripts running.
> 
> @wackfx I haven't published cf support to npm yet so that's why 😉

It installed, but getting back the `ERROR: Could not resolve "*"`, and yes, this is what you get for living in the edge, i was just frustrated, because i was getting errors/problems on non-related things, like pnpm

**Comment by wackfx** [comment]
@JustMrMendez we're all about edges here, like our beloved Cloudflare friends 💛 

Is your problem fixed now at least ? Is it working? You forgot to tell us about this.
Sure I can understand the frustration, believe me when I was debugging this thing yesterday, it was a pain (I don't even use SvelteKit).

But in the end, when it works (and if it works?) - don't you feel relief ? 🙂

**Comment by JustMrMendez** [comment]
> @JustMrMendez we're all about edges here, like our beloved Cloudflare friends 💛 
> 
> Is your problem fixed now at least ? Is it working? You forgot to tell us about this.
> Sure I can understand the frustration, believe me when I was debugging this thing yesterday, it was a pain (and I don't even use SvelteKit).
> 
> But in the end, when it works (and if it works?) - don't you feel relief ? 🙂 

Nope, just removed postgres packages, am trying to make my setup work without it for the moment being, I'll update soon

**Comment by porsager** [comment]
Ok, I got word from Cloudflare it should be exports.worker instead of exports.workerd, so I''ve changed that now.. Let me know if that solves the issues..

**Comment by jahir9991** [comment]
not sure if i am missing anything or not ?
using "@sveltejs/adapter-cloudflare-node": "https://github.com/wackfx/adapter-cloudflare-node",
"postgres": "https://github.com/porsager/postgres",

build failed in prod....
should i need to add anything in vite.config ?



Using @sveltejs/adapter-cloudflare
19:52:17.809 | ✘ [ERROR] Could not resolve "postgres"
19:52:17.809 |  
19:52:17.809 | .svelte-kit/output/server/chunks/hooks.server.js:4:21:
19:52:17.809 | 4 │ import postgres from "postgres";
19:52:17.810 | ╵                      ~~~~~~~~~~
19:52:17.810 |  
19:52:17.810 | The module "./cf/src/index.js" was not found on the file system:
19:52:17.810 |  
19:52:17.810 | node_modules/postgres/package.json:9:14:
19:52:17.810 | 9 │     "worker": "./cf/src/index.js",
19:52:17.811 | ╵               ~~~~~~~~~~~~~~~~~~~
19:52:17.811 |  
19:52:17.811 | You can mark the path "postgres" as external to exclude it from the bundle, which will remove this error.
19:52:17.811

<br class="Apple-interchange-newline">

**Comment by superboss224** [comment]
@porsager @wackfx 
Hello, I've just tested with exports.worker and it now solves the issue, specifying resolve conditions isn't even necessary anymore, it works and now loads from /cf

However, there is new errors now : 
```
18:00:34.662 | ✘ [ERROR] Could not resolve "node:events"
18:00:34.662 |  
18:00:34.663 | node_modules/postgres/cf/polyfills.js:1:29:
18:00:34.663 | 1 │ import { EventEmitter } from 'node:events'
18:00:34.663 | ╵                              ~~~~~~~~~~~~~
18:00:34.663 |  
18:00:34.663 | The package "node:events" wasn't found on the file system but is built into node. Are you trying to bundle for node? You can use "platform: 'node'" to do that, which will remove this error.
18:00:34.663 |  
18:00:34.663 | ✘ [ERROR] Could not resolve "node:buffer"
18:00:34.664 |  
18:00:34.664 | node_modules/postgres/cf/polyfills.js:2:23:
18:00:34.664 | 2 │ import { Buffer } from 'node:buffer'
18:00:34.664 | ╵                        ~~~~~~~~~~~~~
18:00:34.664
```
In the vite.config.js of my SvelteKit project, i'm using:
```
  build: {
		rollupOptions: {
			external: ['cloudflare:sockets', 'cloudflare:events', 'node:stream', 'node:buffer']
		}
	},
```

but it still doesn't resolve the two modules with `node:`

Maybe there's a cloudflare flag to use on the project to load those two modules ? or something else you can do to solve the issue on your side ?

**Comment by wackfx** [comment]
@superboss224 

The answer is in my previous message, you should look at the code I linked.
 
> Can't [override external](https://github.com/sveltejs/kit/blob/8c642cf2d63792e38e6304f840a039d4e4a55781/packages/adapter-cloudflare/index.js#L61). This is important you may want to use node modules available on Pages with the node-compatibility flag. It's missing the node flags we require

esbuild is called by the cloudflare adapter, which doesn't allow for any options. You're only left with either asking them to add the option, or use a custom package like I made.

**Comment by wackfx** [comment]
@jahir9991 
I don't know how you installed the packages, but I believe it's an installation issue, my package should still work as it doesn't override the `worker` usage.

Did you ran the `npm install` I linked ?

**Comment by superboss224** [comment]
@wackfx 
Okay so, [the flag exists](https://developers.cloudflare.com/workers/runtime-apis/nodejs/#:~:text=%E2%80%8B%E2%80%8B%20Enable%20Node.,-js%20from%20the&text=Log%20in%20to%20the%20Cloudflare,your%20Preview%20and%20Production%20deployments.), and `node:buffer` / `node:stream` should work on Cloudflare Pages but esbuild in the cloudflare adapter just don't let those `node:` modules work properly, so it nullifies the flag's effect, and I should ask Cloudflare to take their flag into account into their own adapter to fix the bug

Is that it?

**Comment by wackfx** [comment]
> esbuild in the cloudflare adapter just don't let those node: modules work properly, so it nullifies the flag's effect, and I should ask Cloudflare to take their flag into account into their own adapter to fix the bug

Yes

**Comment by wackfx** [comment]
@superboss224 
> esbuild in the cloudflare adapter just don't let those node: modules work properly, so it nullifies the flag's effect, and I should ask Cloudflare to take their flag into account into their own adapter to fix the bug

Yes. You can have the flag on Cloudflare, if your build system locally doesn't allow it, it won't work.

**Comment by superboss224** [comment]
@wackfx
Okay so, it looks like it's already going to get fixed : 
https://github.com/sveltejs/kit/pull/10544

And he also already has his [temporary driver](https://www.npmjs.com/package/@chientrm/adapter-cloudflare) to "pre-fix" that 

So it should work seamlessly (if there's no other errors) when the pull request is done, all we have to do is wait a little more,
thank you for your help

**Comment by karlhorky** [comment]
ohh nice looks like WinterCG is doing some specification work on `connect()`, maybe it becomes a cross-runtime standard 😮

- https://github.com/vercel/next.js/discussions/46722#discussioncomment-7135863

@porsager what do you think about publishing a new Postgres.js version with the changes in this PR?

**Pull request merged**